src/device_database.rs

```diff
@@ -0,0 +1,627 @@
+
+        Self {
+            id: id.to_string(),
+            sku: sku.to_string(),
+            name: computed_name,
+            room: None,
+            discovered_via: source,
+            first_seen: now,
+            last_seen: now,
+            last_api_sync: None,
+            user_name: None,
+            user_room: None,
+        }
+    }
+
+    /// Create a new device with full API information
+    pub fn new_from_api(
+        id: &str,
+        sku: &str,
+        name: &str,
+        room: Option<&str>,
+        source: DiscoverySource,
+    ) -> Self {
+        let now = Utc::now();
+
+        Self {
+            id: id.to_string(),
+            sku: sku.to_string(),
+            name: name.to_string(),
+            room: room.map(|s| s.to_string()),
+            discovered_via: source,
+            first_seen: now,
+            last_seen: now,
+            last_api_sync: Some(now),
+            user_name: None,
+            user_room: None,
+        }
+    }
+}
+
+/// Compute a device name from SKU and ID (matches Device::computed_name logic)
+fn compute_device_name(sku: &str, id: &str) -> String {
+    // The id is usually "XX:XX:XX:XX:XX:XX:XX:XX" but some devices
+    // report it without colons, and in lowercase. Normalize it.
+    let mut normalized_id = String::new();
+    for c in id.chars() {
+        if c == ':' {
+            continue;
+        }
+        normalized_id.push(c.to_ascii_uppercase());
+    }
+
+    format!(
+        "{}_{}",
+        sku,
+        &normalized_id[normalized_id.len().saturating_sub(4)..]
+    )
+}
+
+impl DeviceDatabase {
+    /// Load database from disk, or create empty if not exists
+    pub fn load(path: &Path) -> anyhow::Result<Self> {
+        if path.exists() {
+            let contents = fs::read_to_string(path)?;
+            let db: DeviceDatabase = serde_json::from_str(&contents)?;
+            log::info!(
+                "Loaded device database with {} devices from {:?}",
+                db.devices.len(),
+                path
+            );
+            Ok(db)
+        } else {
+            log::info!(
+                "No device database found at {:?}, starting fresh",
+                path
+            );
+            Ok(DeviceDatabase::default())
+        }
+    }
+
+    /// Save database atomically (write temp file, then rename)
+    pub fn save(&self, path: &Path) -> anyhow::Result<()> {
+        // Create parent directory if needed
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)?;
+        }
+
+        // Write to temporary file in same directory
+        let temp_path = path.with_extension("json.tmp");
+        let contents = serde_json::to_string_pretty(self)?;
+
+        {
+            let mut file = fs::File::create(&temp_path)?;
+            file.write_all(contents.as_bytes())?;
+            file.sync_all()?; // Ensure data is on disk
+        }
+
+        // Atomic rename
+        fs::rename(&temp_path, path)?;

```

Comment:
re: comments above, if the rename fails for some reason here, `json.tmp` will remain in the filesystem.  Using the tempfile module, the destructor will clean it up for us, so that's another point in its favor.